### DISCUSSION TITLE
NeoLemmix/SuperLemmix: Zooming fails to preserve cursor position relative to level image

### DISCUSSION THREAD

**Discussion by Willicious** [graphics32/graphics32#334] [General] [answered]
Hi Anders,

We have [this bug](https://www.lemmingsforums.net/index.php?topic=6147.msg97831#msg97831) which has been present in NeoLemmix for some time. The easiest way to describe it is that, when increasing the zoom factor, the cursor position does not stay on the same pixel of the zoomed image (but _does_ stay in place on the screen/game window, which is good).

We've found a way to track the pixel that we want to "jump" to when zooming, but using this as a reference during the ChangeZoom procedure would require significant refactoring of various processes and state initializations that happen when first preparing the game for rendering; not ideal.

We were wondering - does GR32 have an existing way to "move" the image such that the pixel under the cursor in zoom factor 1 (for example) stays the same when increasing to zoom factor 2 (for example)? At present, the image jumps some distance to the left or right when zooming in and out.

We're looking specifically at OffsetHorz, which can be used to nudge the image, but we need to know by how much. Hopefully, GR32 has existing ways of calculating this, ideally without callbacks.

Any suggestions you might have will be much appreciated.

Kind regards,

Will

**Comment by andersmelander** [comment, marked as answer]
Hi,

I feel your pain; The scale/offset stuff can be pretty confusing. Also for me :-)

> does GR32 have an existing way to "move" the image such that the pixel under the cursor in zoom factor 1 (for example) stays the same when increasing to zoom factor 2 (for example)?

I guess you can just use the `TCustomImage32.Zoom` method:

https://github.com/graphics32/graphics32/blob/445d0212e1f49b349ef6b8760c0ea64b2906c9ee/Source/GR32_Image.pas#L617

The `APivot` parameter specifies the position, in bitmap coordinates, that should stay fixed.

Assuming you're zooming in a MouseDown event handler, so your coordinates are relative to the control, then you can use [`TCustomImage32.ControlToBitmap`](https://github.com/graphics32/graphics32/blob/445d0212e1f49b349ef6b8760c0ea64b2906c9ee/Source/GR32_Image.pas#L593) to get the bitmap coordinate for the `APivot` parameter.

You might want to look at the [HoverZoom example](https://github.com/graphics32/graphics32/tree/master/Examples/General/HoverZoom) which demonstrate some different ways of zooming while maintaining the "pivot point".


Actually, now that I look at your code in `TGameWindow.ChangeZoom` I can see that all your additional constraints on the viewport size/position will conflict with what Zoom does. I'm not sure how much of it your really need.
FWIW, the following works for me:

```pascal
procedure TGameWindow.ChangeZoom(aNewZoom: Integer; NoRedraw: Boolean = false);
var
  Pivot: TPoint;
begin
  aNewZoom := Max(Min(fMaxZoom, aNewZoom), 1);
  if (aNewZoom = fInternalZoom) and not NoRedraw then
    Exit;

  SkillPanel.Image.BeginUpdate;
  try
    Pivot := Img.ScreenToClient(Mouse.CursorPos);
    Pivot := Img.ControlToBitmap(Pivot);
    Img.Zoom(aNewZoom, Pivot);

    fInternalZoom := aNewZoom;

    ApplyResize(False);

    SetRedraw(rdRedraw);
  finally
    SkillPanel.Image.EndUpdate;
  end;
end;
```

Don't bother with the animated zoom; It doesn't play well with how you create the image.

**Reply by Willicious** [reply]
Thanks for the code! I get an undeclared identifier error on `Img.Zoom`. Here are the GR32 declarations in uses, perhaps a specific one is missing:

`GR32, GR32_Image, GR32_Layers, GR32_Resamplers`

**Reply by andersmelander** [reply]
What version (I need the commit SHA) of Graphics32 are you using?

**Reply by Willicious** [reply]
I'm on bb52dd9b, I'll try updating to the latest commit

**Reply by Willicious** [reply]
Done! The new ChangeZoom procedure works, thanks so much for finding a fix! :)

**Reply by andersmelander** [reply]
No that revision should be fine.
Can you verify that your TImage32 has a Zoom method (two of them actually):

https://github.com/graphics32/graphics32/blob/bb52dd9b66622d8995a4cf98d33c8ba9f5e8ed89/Source/GR32_Image.pas#L597-L598

**Reply by andersmelander** [reply]
Okay. That's fine then.

**Reply by Willicious** [reply]
Thanks again. Confirmed, TImage32 has both Zoom methods as demonstrated.

The only issue I can find with it is that the cursor is repositioned when zooming between factor 1 and factor 2 (this is most pronounced at the extreme level edges/corners). Between factor 2 and higher, cursor position is preserved perfectly. I'm not sure exactly what's going on here, but it's definitely a promising start.

For clarity, in this context "factor" is defined by GameWindow's `fInternalZoom` field - this may not be quite the same value as the GR32 definition (actually, if the values are different for any reason, that could be the source of this bug).

**Reply by andersmelander** [reply]
I can see that you've committed the change as-is to the master branch (that's brave :-) ). I'll take another look at it later tonight.

**Reply by andersmelander** [reply]
First of all, the code I posted was not meant to be the final solution to your problem. It was just an example of how to use the Zoom method in the context of your code.

Anyway, there are a couple of problems:

1. After calling `Zoom` you then call `TGameWindow.ApplyResize`.
2. `TGameWindow.ApplyResize` resizes and repositions the `TImage32` control which completely invalidates the zoom/pivot stuff we've just done since it moves the viewport relative to the screen/cursor.
3. `TGameWindow.ApplyResize`  also (optionally) adjusts `TImage32.OffsetHorz` and `OffsetVert` - which again makes the pivot stuff pointless.

I don't know what the overall purpose of `ApplyResize` is but it would probably be better to just `alClient` align the image control and then leave it be.

I have seen it suggested that the skill panel be made a layer but I don't see any benefit or purpose of that since it never needs to be blended on the main game screen. Parts of the panel could be layers but that's another story.

So it seems you're not quite there.

**Reply by Willicious** [reply]
> the code I posted was not meant to be the final solution to your problem
>

Sure, but it works (at least partially) and it's a big step in the right direction to what we had previously, so I'm happy to go ahead with it. I realise I probably should have branched rather than committed to master, but I'm still not used to working with branches. Ironically, using branches seems far more "brave" to me than committing to master!

Next time, I'll use branches, promise! :)
 
> 1. After calling `Zoom` you then call `TGameWindow.ApplyResize`.
> 2. `TGameWindow.ApplyResize` resizes and repositions the `TImage32` control which completely invalidates the zoom/pivot stuff we've just done since it moves the viewport relative to the screen/cursor.
> 3. `TGameWindow.ApplyResize`  also (optionally) adjusts `TImage32.OffsetHorz` and `OffsetVert` - which again makes the pivot stuff pointless.
> 

If that's the case, perhaps the ApplyResize call can just be removed?

> I don't know what the overall purpose of `ApplyResize` is but it would probably be better to just `alClient` align the image control and then leave it be.
> 

I'm not sure how to do this yet, but at this point I'm willing to try anything you think might be best. I'll do some reading of the GR32 documentation later and see if I can figure this one out myself.

Thanks again for your help so far.

**Reply by andersmelander** [reply]
> Ironically, using branches seems far more "brave" to me than committing to master!

Since you are working with public code branches mean that you don't have to worry about what consequences your experiments have on other people or if it will ever amount to anything.
You can also use branches that only exist locally and only push them to the server once you have a need for it. In Graphics32 I currently have something like 15-20 local branches with various stuff I'm working on. Some of it may never amount to anything and will never get merged and some of it needs to be cleaned up before I can have anyone else see it.

> perhaps the ApplyResize call can just be removed?

Not entirely. It should at least be replaced with something else because it seems that it is at least responsible for sizing and positioning the two main images.

>  I'll do some reading of the GR32 documentation later and see if I can figure this one out myself.

The GR32 documentation is very much out of date so I'm not sure that will do you much good. I also think the main problem here is that there is no clear principle in how the NL game screen is structured. That's how it is with old code. It's one patch on top of the other, kept together with chewing gum and rubber bands.

**Reply by Willicious** [reply]
Re: TGameWindow.ApplyResize
 
> Not entirely. It should at least be replaced with something else because it seems that it is at least responsible for sizing and positioning the two main images.
> 

I wonder why the problem only occurs between zoom levels 1 and 2...

Between 2 and 3, 3 and 4, etc, the pivot point is preserved. ChangeZoom is called with each increase/decrease in zoom factor, and thus so is ApplyResize. This makes it somewhat difficult to understand how that particular procedure could be causing the problem, even if we agree that it might need a bit of refactoring/rewriting.

**Reply by andersmelander** [reply]
I think the observed behavior depends on the size (resolution, actually) of the screen but I don't know enough about how NL scales its graphics to say for sure.

It is only possible to maintain the pivot if the viewport can be panned and from what I can see in NL it can only be panned when the gamescreen is larger than the screen. I think that is a reasonable constraint.

On my system the gamescreen becomes larger than the screen at zoomlevel 3 and I'm observing the inability to maintain pivot at zoomlevels 1-2-3. Between 2 and 3 the pivot is **only** maintained if that would not conflict with NL's pan constraints, so that is usually a point near the center.

So after having played with it a bit I think the current behavior is actually the best you can hope for.

**Reply by Willicious** [reply]
> from what I can see in NL it can only be panned when the gamescreen is larger than the screen.
> 

Can you point to which part of the code specifies this, or is it more of a general outcome/symptom resulting from more widespread logic?

If the former, I think it's worth at least experimenting with. If the latter, I'll probably just leave it as it is.

**Reply by andersmelander** [reply]
I would think that the pan constraint is by-design with the goal of keeping as much of the gamescreen as possible visible on the screen.

Why would you _not_ want that constraint?

**Reply by Willicious** [reply]
> I would think that the pan constraint is by-design with the goal of keeping as much of the gamescreen as possible visible on the screen.
> 
> Why would you _not_ want that constraint?

I would want it for all scenarios _except_ when zooming in. So, if we can find the part of the code that says "keep as much of the gamescreen as possible visible", we can add "except when zooming in, in which case listen only to the pivot point".

**Reply by andersmelander** [reply]
I guess you have your reasons but I don't think that makes any sense and I think you will break other parts of the UI if you manage to tweak the logic. For example the minimap displays the viewport and I doubt that it was made to handle the case where the viewport exceeds the gamescreen.

**Reply by Willicious** [reply]
To be fair, even Lix (Simon's Lemmings clone) has some auto-panning when zooming in at the edges from the lowest zoom factor. Only when it reaches some higher factor (probably where the gamescreen is larger than the display screen, as you've noted in NL) does it begin to use a specific pivot.

So, it could be that the nature of the two display elements "keep everything as visible as possible" and "zoom in to a very specific point under the mouse cursor" are mutually incompatible below a certain display size:zoom factor, whatever the system's logic.

I'll close this one out as resolved. Your solution has definitely improved things, and seems to integrate well into the existing logic.

**Reply by andersmelander** [reply]
I have a Lemmix bugfix for an unrelated issue. If I should create a pull request with it, which of your repos should I make it against? The Github or the Bitbucket one?

The fix applies to Neolemmix, since that is the version I can run (I don't have the Superlemmix assets), but the offending code is in Superlemmix too so I assume that it has the same problem. I will describe the nature of the problem in the pull request.

**Reply by Willicious** [reply]
Thanks, it's appreciated.

Either is fine. The GitHub is a remote, the Bitbucket is the "original", I tend to commit there first and then push to GitHub once I'm happy I won't make any other changes.
